pmd-core/src/main/java/net/sourceforge/pmd/lang/metrics/ParameterizedMetricKey.java

```diff
@@ -57,13 +59,22 @@ private static int code(MetricKey key, MetricVersion version) {
     }
 
 
-    /** Builds a parameterized metric key. */
-    public static ParameterizedMetricKey getInstance(MetricKey<? extends Node> key, MetricVersion version) {
+    /**
+     * Builds a parameterized metric key.
+     *
+     * @param key     The key
+     * @param version The version
+     * @param <N>     The type of node of the metrickey
+     *
+     * @return An instance of parameterized metric key corresponding to the parameters
+     */
+    @SuppressWarnings("unchecked")
+    public static <N extends Node> ParameterizedMetricKey<N> getInstance(MetricKey<N> key, MetricVersion version) {
         int code = code(key, version);
```

Comment:
Just seeing this line - this might be a theoretical problem: the code (hashCode) might not be unique and we might get collisions. Not sure, if this is a real problem, but in theory, it exists.

Assuming, that we only pool this class not for performance reasons, but to ensure, only one instance exists (so we can compare two ParametrizedMetricKeys later with a simple `==`): We could use a HashMap<ParametrizedMetricKey, ParameterizedMetricKey>. We would always are (temporarily) a new instance of the ParameterizedMetricKey, but return the one on the hashMap, if it exists...
Then the HashMap takes care of potential collisions.